d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ZKFailoutConfigProvider.java

```diff
@@ -90,6 +91,10 @@ public FailoutConfig getFailoutConfig(String clusterName)
     return failedoutClusterManager != null ? failedoutClusterManager.getFailoutConfig() : null;
   }
 
+  public Set<String> getClusters() {
+    return _failedoutClusterManagers.keySet();
```

Comment:
Should we make a copy [given the doc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#keySet--)? Wondering if it will give an exception when the underlying map is modified while we are iterating the set.

Reply:
Done. The docs say that the iterator will not throw an exception but reflects at a state at the time of creation or in the future and are weakly consistent. For a metrics use case I think this is acceptable (we have that issue reading each config over time as well).